nixos/modules/security/pam.nix

```diff
@@ -369,233 +409,272 @@ let
+        }) ++ (optional config.krb5.enable {
+          control = "optional"; path = "${pam_krb5}/lib/security/pam_krb5.so";
+        }) ++ (optional cfg.otpwAuth {
+          control = "optional"; path = "${pkgs.otpw}/lib/security/pam_otpw.so";
+        }) ++ (optional cfg.startSession {
+          control = "optional"; path = "${pkgs.systemd}/lib/security/pam_systemd.so";
+        }) ++ (optional cfg.forwardXAuth {
+          control = "optional"; path = "pam_xauth.so"; arguments = ["xauthpath=${pkgs.xorg.xauth}/bin/xauth" "systemuser=99"];
+        }) ++ (optional (cfg.limits != []) {
+          control = "required"; path = "${pkgs.pam}/lib/security/pam_limits.so"; arguments = ["conf=${makeLimitsConf cfg.limits}"];
+        }) ++ (optional (cfg.showMotd && config.users.motd != null) {
+          control = "optional"; path = "${pkgs.pam}/lib/security/pam_motd.so"; arguments = ["motd=${motd}"];
+        }) ++ (optional (cfg.enableAppArmor && config.security.apparmor.enable) {
+          control = "optional"; path = "${pkgs.apparmor-pam}/lib/security/pam_apparmor.so"; arguments = ["order=user,group,default" "debug"];
+        }) ++ (optional (cfg.enableKwallet) ({
+          control = "optional"; path = "${pkgs.plasma5.kwallet-pam}/lib/security/pam_kwallet5.so";
+          arguments = "kwalletd=${pkgs.kdeFrameworks.kwallet.bin}/bin/kwalletd5";
+        })) ++ (optional (cfg.enableGnomeKeyring) {
+          control = "optional"; path = "${pkgs.gnome3.gnome-keyring}/lib/security/pam_gnome_keyring.so"; arguments = ["auto_start"];
+        }) ++ (optional cfg.gnupg.enable {
+          control = "optional"; path = "${pkgs.pam_gnupg}/lib/security/pam_gnupg.so";
+          arguments = optional cfg.gnupg.noAutostart "no-autostart";
+        }) ++ (optional (config.virtualisation.lxc.lxcfs.enable) {
+          control = "optional"; path = "${pkgs.lxc}/lib/security/pam_cgfs.so"; arguments = [ "-c all" ];
+        });
+      };
     };
-
   };
 
-
   inherit (pkgs) pam_krb5 pam_ccreds;
 
   use_ldap = (config.users.ldap.enable && config.users.ldap.loginPam);
   pam_ldap = if config.users.ldap.daemon.enable then pkgs.nss_pam_ldapd else pkgs.pam_ldap;
 
   # Create a limits.conf(5) file.
-  makeLimitsConf = limits:
-    pkgs.writeText "limits.conf"
-       (concatMapStrings ({ domain, type, item, value }:
-         "${domain} ${type} ${item} ${toString value}\n")
-         limits);
+  makeLimitsConf = limits: pkgs.writeText "limits.conf" (
+    concatMapStrings ({ domain, type, item, value }:
+      "${domain} ${type} ${item} ${toString value}\n"
+    ) limits
+  );
 
   motd = pkgs.writeText "motd" config.users.motd;
 
-  makePAMService = name: service:
-    { name = "pam.d/${name}";
-      value.source = pkgs.writeText "${name}.pam" service.text;
-    };
-
-in
-
-{
+  makePAMService = name: service: let
+    argumentToString = value: if hasInfix " " value then "[${escape [ "]" ] value}]" else value;
+    controlToString = value:
```

Comment:
It would be nice to use the apply arg to `mkOption` to achieve this

Reply:
Something like this:
```
      control = mkOption {
        type = controlType;                                                                    
        apply = value: if isString value then value else
                       concatStringsSep " " (mapAttrsToList (n: v: "${name}=${toString value}"));                      
      };                                                   
      args = mkOption {
        default = [];
        apply = value: concatStringsSep " " (map toString value);
      };
```

Reply:
See
https://github.com/NixOS/nixpkgs/pull/105319/files#diff-b08ed0eadd5d6b3a3da3a911d502ff1de42cc47bc4bd0c296576a247fd4ac96fR25-R27
and
https://github.com/NixOS/nixpkgs/pull/105319/files#diff-b08ed0eadd5d6b3a3da3a911d502ff1de42cc47bc4bd0c296576a247fd4ac96fR38